src/features/surveys/rpc/getSurveyResponseStats.ts

```diff
@@ -0,0 +1,251 @@
+  questions: QuestionStats[];
+  submissionStats: SubmissionStats[];
+};
+
+export const getSurveyResponseStatsDef = {
+  handler: handle,
+  name: 'getSurveyResponseStats',
+  schema: paramsSchema,
+};
+
+type ResponseStatsCounter = {
+  answerCounter: number;
+  selectedOptions: Record<
+    number,
+    {
+      count: 0;
+      option: ZetkinSurveyOption;
+    }
+  >;
+  topWordFrequencies: Record<string, number>;
+  totalSelectedOptionsCounts: number;
+  totalUniqueWordCount: number;
+  totalWordCounts: number;
+  wordFrequencies: Record<string, number>;
+};
+
+export default makeRPCDef<Params, SurveyResponseStats>(
+  getSurveyResponseStatsDef.name
+);
+
+async function handle(
+  params: Params,
+  apiClient: IApiClient
+): Promise<SurveyResponseStats> {
+  const { orgId, surveyId } = params;
+
+  const [survey, submissions]: [
+    ZetkinSurveyExtended,
+    ZetkinSurveySubmission[]
+  ] = await Promise.all([
+    apiClient.get<ZetkinSurveyExtended>(
+      `/api/orgs/${orgId}/surveys/${surveyId}`
+    ),
+    apiClient.get<ZetkinSurveySubmission[]>(
+      `/api/orgs/${orgId}/surveys/${surveyId}/submissions`
+    ),
+  ]);
+
+  const responses: (ZetkinSurveySubmission | undefined)[] = await Promise.all(
+    submissions.map((submission) =>
+      apiClient
+        .get<ZetkinSurveySubmission>(
+          `/api/orgs/${orgId}/survey_submissions/${submission.id}`
+        )
+        .catch(() => Promise.resolve(undefined))
+    )
+  );
+
+  const responseStatsCounters: Record<number, ResponseStatsCounter> = {};
+
+  survey.elements.forEach((question) => {
+    if (question.type === ELEMENT_TYPE.TEXT) {
+      return;
+    }
+
+    // initialize counters for each question
+    responseStatsCounters[question.id] = {
+      answerCounter: 0,
+      selectedOptions: {},
+      topWordFrequencies: {},
+      totalSelectedOptionsCounts: 0,
+      totalUniqueWordCount: 0,
+      totalWordCounts: 0,
+      wordFrequencies: {},
+    };
+    // handle questions with options as response
+    if (
+      question.question.response_type === RESPONSE_TYPE.TEXT ||
+      !question.question.options
+    ) {
+      return;
+    }
+    (question as ZetkinSurveyOptionsQuestionElement).question.options!.forEach(
+      (option) => {
+        responseStatsCounters[question.id].selectedOptions[option.id] = {
+          count: 0,
+          option: option,
+        };
+      }
+    );
+  });
+
+  // this handles questions with text as response
+  responses.forEach((response) => {
+    if (!response || !response.responses) {
+      return;
+    }
+
+    response.responses.forEach((response) => {
+      if (
+        !Object.prototype.hasOwnProperty.call(
+          responseStatsCounters,
+          response.question_id
+        )
+      ) {
+        return;
+      }
+
+      if (!('response' in response) || response.response.length > 0) {
+        responseStatsCounters[response.question_id].answerCounter++;
+      }
+
+      if ('response' in response) {
```

Comment:
I read these uses of `in` as being a way to assess the type of a `response`. If that's indeed the case, I think it would be better to have [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates), which are much more self-documenting.

That way, the code could look like this instead:

```suggestion
      if (isTextResponse(response)) {
```